bika/lims/exportimport/instruments/generic/two_dimension.py

```diff
@@ -93,29 +93,63 @@ def is_keyword(kw):
     return len(bsc(getKeyword=kw))
 
 
-def find_kw(ar_or_sample, kw):
+def find_analyses(ar_or_sample):
     """ This function is used to find keywords that are not on the analysis
         but keywords that are on the interim fields.
 
         This function and is is_keyword function should probably be in
         resultsimport.py or somewhere central where it can be used by other
         instrument interfaces.
     """
-    keyword = None
     bc = api.get_tool(CATALOG_ANALYSIS_REQUEST_LISTING)
     ar = bc(portal_type='AnalysisRequest', id=ar_or_sample)
     if len(ar) == 0:
         ar = bc(portal_type='AnalysisRequest', getSampleID=ar_or_sample)
     if len(ar) == 1:
         obj = ar[0].getObject()
         analyses = obj.getAnalyses(full_objects=True)
-        for analysis in analyses:
-            interims = hasattr(analysis, 'getInterimFields') \
-                        and analysis.getInterimFields() or []
-            for interim in interims:
-                if interim['keyword'] == kw:
-                    keyword = analysis.getKeyword()
-                    break
+        return analyses
+    return []
+
+
+def find_analysis_interims(ar_or_sample):
+    """ This function is used to find keywords that are not on the analysis
+        but keywords that are on the interim fields.
+
+        This function and is is_keyword function should probably be in
+        resultsimport.py or somewhere central where it can be used by other
+        instrument interfaces.
+    """
+    interim_fields = []
+    analyses = find_analyses(ar_or_sample)
+    for analysis in analyses:
+        interims = []
+        if hasattr(analysis, 'getInterimFields'):
+            interims = analysis.getInterimFields()
+        for interim in interims:
+            if interim['keyword'] not in interim_fields:
+                interim_fields.append(interim['keyword'])
+    return interim_fields
+
+
+def find_kw(ar_or_sample, kw):
+    """ This function is used to find keywords that are not on the analysis
+        but keywords that are on the interim fields.
+
+        This function and is is_keyword function should probably be in
+        resultsimport.py or somewhere central where it can be used by other
+        instrument interfaces.
+    """
+    keyword = None
+    analyses = find_analyses(ar_or_sample)
+    for analysis in analyses:
+        interims = []
+        if hasattr(analysis, 'getInterimFields'):
+            interims = analysis.getInterimFields()
+        for interim in interims:
+            if interim['keyword'] == kw:
+                keyword = analysis.getKeyword()
+                break
```

Comment:
Note that you force a break only from inside the inner loop, but the outer loop (analyses) will continue unnecessarily.
The complexity of this function can be reduced as follows:

```python
def find_kw(ar_or_sample, kw):
    for analysis in find_analyses(ar_or_sample):
        interims = api.safe_getattr(analysis, 'getInterimFields', default=[])
        keywords = map(lambda item['keyword'], interims)
        if kw in keywords:
            return analysis.getKeyword()
    return None
```
Given the fact that the two first lines from within the loop are duplicated in `find_analysis_interims` I'd also suggest to create a function `get_interims_keywords`, so the complexity could be reduced a bit more::

```python
def get_interims_keywords(analysis):
    keywords = api.safe_get_attr(analysis, 'getInterimFields', default=[])
    return map(lambda item['keyword'], interims)

def find_analysis_interims(ar_or_sample):
    for analysis in find_analyses(ar_or_sample):
        keywords = get_interims_keywords(analysis)
        interim_fields.extend(keywords)
    return list(set(interim_fields))

def find_kw(ar_or_sample, kw):
    for analysis in find_analyses(ar_or_sample):
        if kw in get_interims_keywords(analysis):
            return analysis.getKeyword()
    return None
```

Reply:
Done